docs/adr/0001-update-logins-api.md

```diff
@@ -0,0 +1,104 @@
+* Option 2 - Implement a different key management approach
+
+
+## Decision Outcome
+
+(I'm purposely leaving this blank until we arrive at a decision.)
+
+
+## Pros and Cons of the Options
+
+### Option 1 - Reduce the API functions that require the encryption key and pass the key to the remaining functions
+
+* Description
+
+  Currently the below logins API functions would require the per-field encryption key:
+    - [add](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L362)
+    - [update](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L611)
+    - [get_all](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L273)
+    - [get_by_base_domain](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L279)
+    - [get_by_id](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L330)
+    - [check_valid_with_no_dupes](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L663)
+    - [potential_dupes_ignoring_username](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L721)
+    - [import_multiple](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L455)
+
+      **Note:** Functions related to sync have been omitted as it is assumed they will have access to decrypted data.
+
+  Propsed changes:
+    - Combine the `add` and `update` functions into a new `add_or_update` function
+      - This will allow the removal of consumer code that distinguishes when a login record should be created or updated
+    - Pass the per-field encryption key to the `import_multiple` function
+      - This function will be removed once the Fennec to Fenix migration period ends
+    - Remove both the `potential_dupes_ignoring_username` and `check_valid_with_no_dupes` from the API
+      - Neither function is called in Firefox iOS
+      - Android Components uses both to provide validation and de-duping before logins are added or updated so we can eliminate the need to externalize these functions by replicating this logic in the new `add_or_update` function
+    - Create a `decrypt_and_fixup_login` function that both decrypts a login _and_ performs the validate and fixup logic
+      - This will eliminate the need for the `get_all`, `get_by_base_domain`, and `get_by_id` API functions to perform the fixup logic
+
+    Making the above changes will reduce the API functions requiring the encryption key to the following:
+    - `add_or_update`
+    - `decrypt_and_fixup_login`
+    - `import_multiple`
```

Comment:
Another change is how the username/password fields on the Login object works.  It seems minor, but there's definitely some edge cases.  I think the main questions are:

  - How do you get/set values the fields?
  - How do you set the username field to missing?
  - How do we know when the time_password_changed field should be bumped?

Reply:
@bendk Sorry for the delayed response, I somehow missed this.


>     * How do you get/set values the fields?
What edge cases were you thinking of specifically. I'm assuming you're referring to temporarily holding the the decrypted username of password values to make the deduping or fixup work cleaner. If that's the case, I assumed we might need internal structs however I'll readily admit to not thinking too much about this as this doc is a bit higher level. Were there specific cases you were concerned about?

>     * How do you set the username field to missing?
So we've already set the precedent that an empty string value indicates that the record has been scrubbed, which is a meaningful distinction that we probably don't want to overload. So maybe a missing username would be NULL value in the database and perhaps a None Option in the Logins struct. @mhammond How are you feeling about `Option<String>` usernames? 😬


>     * How do we know when the time_password_changed field should be bumped?

At present this only happens during db updates. So in the new `add_or_update` field I'm thinking we'd continue updating that field if we determine that the login already exists in the database. Since that function will have the encryption key, this shouldn't be a problem. Let me know if there are edge cases here that you're thinking of that I may not have considered.

Reply:
All the above sounds fine, but as a meta-comment, I don't think it's necessary for this ADR to get into that level of detail as that doesn't really drive the higher-level decision about what scheme to use. (That's also *partially* true for some other points, but there is a fine line because we do want to thrash out things not *directly* related but that still might cause problems with the scheme we agree on)

Reply:
Part of the question was really aim aimed at lower-level details that I agree the ADR doesn't need to address.  The one thing that I'd like to see expanded on is the general flow of fetch login -> decrypt login -> save login back to the db.

What `decrypt_and_fixup()` do?
  - Adds plaintext username/password data to the existing dictionary
  - Returns a new dictionary with the plaintext data instead of encrypted data (converts `EncryptedLogin` to `Login`)

My suggestion is that `decrypt_and_fixup()` returns a new type.  Otherwise we can get into ambiguous situations, like if the decrypted password and encrypted password fields don't match up.

What get passed to `add_or_update()`
  - The dictionary that was fetched from the DB
  - A new dictionary that the consumer code creates
  - Is the data encrypted or plaintext?

I kind of think that `add_or_update()` should input a newly created dictionary with plaintext data, but I'm not sure.